### PR TITLE
Add a shading example to the rest client docs

### PR DESCRIPTION
**Pull request by hub-cap**
This adds a maven and gradle example for shading as well as the caveat
about shading and commons logging.

### CODE REVIEW

**Comment by hub-cap** [comment]
I validated the change using the `build_docs.pl --doc docs/java-rest/index.asciidoc`

File: docs/java-rest/low-level/usage.asciidoc [hunk 1]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
 - commons-codec:commons-codec
 - commons-logging:commons-logging
 
+[[java-rest-low-usage-shading]]
+=== Shading
+
+In order to avoid version conflicts, the dependencies can be shaded and packaged
+within the client in a single JAR file (sometimes called "uber jar" or "fat
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `sometimes called "uber jar"` -> `sometimes called an "uber jar"`

File: docs/java-rest/low-level/usage.asciidoc [hunk 2]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
...
+jar"). Shading a dependency consists of taking its content (resources files and
+Java class files), rename its dependency packages before putting them in the
+same JAR file as the low-level Java REST client. Shading a jar can be
+accomplished by 3rd party plugins for gradle and maven.
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `gradle` -> `Gradle`

Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `maven` -> `Maven`

Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `3rd party` -> `3rd-party`

File: docs/java-rest/low-level/usage.asciidoc [hunk 3]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
 - commons-codec:commons-codec
 - commons-logging:commons-logging
 
+[[java-rest-low-usage-shading]]
+=== Shading
+
+In order to avoid version conflicts, the dependencies can be shaded and packaged
+within the client in a single JAR file (sometimes called "uber jar" or "fat
+jar"). Shading a dependency consists of taking its content (resources files and
+Java class files), rename its dependency packages before putting them in the
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `, rename its dependency packages` -> ` and renaming some of its packages`

File: docs/java-rest/low-level/usage.asciidoc [hunk 4]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
...
+same JAR file as the low-level Java REST client. Shading a jar can be
+accomplished by 3rd party plugins for gradle and maven.
+
+Be advised that shading a jar also has implications. Shading the commons logging
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `commons logging` -> `Commons Logging`

File: docs/java-rest/low-level/usage.asciidoc (hunk 1 above)
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: You use `jar` and `JAR` throughout, let's settle on one (how about `JAR`?).

File: docs/java-rest/low-level/usage.asciidoc [hunk 5]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
...
+layer, for instance, means that 3rd party logging backends need to be shaded as
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `3rd party` -> `3rd-party`

File: docs/java-rest/low-level/usage.asciidoc [hunk 6]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
...
+well.
+
+[[java-rest-low-usage-shading-maven]]
+==== Maven Sample
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: I wonder if this should `float` so it's on the same page as the remainder of the content?

File: docs/java-rest/low-level/usage.asciidoc [hunk 7]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
...
+                    </configuration>
+                </execution>
+            </executions>
+        </plugin>
+    </plugins>
+</build>
+--------------------------------------------------
+
+[[java-rest-low-usage-shading-gradle]]
+==== Gradle configuration
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: I wonder if this should `float` so it's on the same page as the remainder of the content?

File: docs/java-rest/low-level/usage.asciidoc (hunk 6 above)
Comment (jasontedor) [reply docs/java-rest/low-level/usage.asciidoc]: One example is `Maven Sample` and the other is `Gradle configuration`; pick a common term?

File: docs/java-rest/low-level/usage.asciidoc [hunk 8]
```diff
@@ -65,6 +65,82 @@ http://hc.apache.org/httpcomponents-asyncclient-dev/[Apache Http Async Client]
...
+accomplished by 3rd party plugins for gradle and maven.
+
+Be advised that shading a jar also has implications. Shading the commons logging
+layer, for instance, means that 3rd party logging backends need to be shaded as
+well.
+
+[[java-rest-low-usage-shading-maven]]
+==== Maven Sample
+
+Here is a configuration using the maven
```
Comment (jasontedor) [inline docs/java-rest/low-level/usage.asciidoc]: `maven` -> `Maven`

**Review by jasontedor — CHANGES_REQUESTED** [review]
It looks great, I left some nits.

**Review by dadoonet — COMMENTED** [review]
I wonder if we would like to add a link to https://www.elastic.co/blog/to-shade-or-not-to-shade
In this post, I was providing a sample project available at: https://github.com/dadoonet/shaded-test

Might not be a good idea to reference a blog post from the reference guide though.

**Comment by hub-cap** [comment]
Ty @jasontedor.

@dadoonet yea I was hesitant to link 3rd party resources (even our blog) that weren't under our repos version control.

File: docs/java-rest/low-level/usage.asciidoc (hunk 6 above)
Comment (hub-cap) [reply docs/java-rest/low-level/usage.asciidoc]: WRT `float`, It looks like 4-deep `====` does not create pages, an example is right above this in the `Maven` section, 

https://www.elastic.co/guide/en/elasticsearch/client/java-rest/5.6/java-rest-low-usage-maven.html

You do, however *have* to `[float]` anything thats 3 `===` deep to keep it on the same rendered page.

File: docs/java-rest/low-level/usage.asciidoc (hunk 7 above)
Comment (hub-cap) [reply docs/java-rest/low-level/usage.asciidoc]: see https://github.com/elastic/elasticsearch/pull/26449#discussion_r136418174

**Comment by hub-cap** [comment]
fixed up and ready to roll @jasontedor

File: docs/java-rest/low-level/usage.asciidoc (hunk 6 above)
Comment (jasontedor) [reply docs/java-rest/low-level/usage.asciidoc]: Good catch, thanks for the clarification.

**Review by jasontedor — APPROVED** [review]

**Comment by jasontedor** [comment]
Thanks @hub-cap!

**Pull request merged**